File: game_sim/roster.py
import json

class Player:
    def __init__(self, data):
        self.name = data["name"]
        self.position = data["position"]
        self.speed = data.get("speed", 50)
        self.strength = data.get("strength", 50)
        self.intelligence = data.get("intelligence", 50)
        self.endurance = data.get("endurance", 50)
        self.fatigue = data.get("fatigue", 0)
        self.in_game = data.get("in_game", False)
        self.stats = data.get("stats", {})

        # Skill-specific
        self.passing = data.get("passing")
        self.decision_making = data.get("decision_making")
        self.elusiveness = data.get("elusiveness")
        self.vision = data.get("vision")
        self.hands = data.get("hands")
        self.route_running = data.get("route_running")
        self.run_blocking = data.get("run_blocking")
        self.pass_blocking = data.get("pass_blocking")
        self.rushing = data.get("rushing")
        self.tackling = data.get("tackling")
        self.coverage = data.get("coverage")
        self.kick_power = data.get("kick_power")
        self.kick_accuracy = data.get("kick_accuracy")
        self.punt_power = data.get("punt_power")
        self.punt_accuracy = data.get("punt_accuracy")

        self._original_attrs = {
            "speed": self.speed,
            "strength": self.strength,
            "elusiveness": self.elusiveness,
            "vision": self.vision,
            "hands": self.hands,
            "route_running": self.route_running,
            "run_blocking": self.run_blocking,
            "pass_blocking": self.pass_blocking,
            "rushing": self.rushing,
            "tackling": self.tackling,
            "coverage": self.coverage
        }

    def is_offense(self):
        return self.position in {"QB", "RB", "WR", "TE", "OL", "K", "P"}

    def is_defense(self):
        return self.position in {"DL", "ROLB", "MLB", "LOLB", "CB", "S"}

    def fatigue_level(self):
        return self.fatigue / max(self.endurance, 1)

    def to_dict(self):
        return self.__dict__.copy()


class Team:
    def __init__(self, name, offense, defense):
        self.name = name
        self.offense = [Player(p) for p in offense]
        self.defense = [Player(p) for p in defense]

    def get_offense(self, side="offense"):
        players = getattr(self, side)
        return [p for p in players if p.in_game]
    
    def get_all_offense(self, side="offense"):
        players = getattr(self, side)
        return [p for p in players]
    
    def get_defense(self, side="defense"):
        players = getattr(self, side)
        return [p for p in players if p.in_game]
        
    def get_all_defense(self, side="defense"):
        players = getattr(self, side)
        return [p for p in players]

    def get_player_by_name(self, name):
        for p in self.offense + self.defense:
            if p.name == name:
                return p
        return None
    
    def recover_bench_players(self):
        for p in self.offense + self.defense:
            if not p.in_game:
                recovery = p.endurance / 10
                p.fatigue = max(0, p.fatigue - recovery)

    def sub_skill_position_players(self, fatigue_threshold=50):
        from operator import itemgetter

        skill_positions = ['RB', 'WR', 'TE']

        for pos in skill_positions:
            # Get current in-game players of this position
            starters = [p for p in self.offense if p.position == pos and p.in_game and p.fatigue > fatigue_threshold]
            if not starters:
                continue  # no one too tired

            # Get bench players of same position who are fresh enough
            bench = [p for p in self.offense if p.position == pos and not p.in_game and p.fatigue <= fatigue_threshold]
            if not bench:
                continue  # no fresh players available

            # Prioritize fresh subs by fatigue ASC, then by total stats DESC
            def stat_sum(player):
                return sum(player.stats.values())

            bench.sort(key=lambda p: (-stat_sum(p), p.fatigue))

            # Sub one-for-one
            for tired_player in starters:
                if not bench:
                    break
                sub = bench.pop(0)
                tired_player.in_game = False
                sub.in_game = True

    def sub_defensive_players(self, fatigue_threshold=50):
        positions = set(p.position for p in self.defense)

        for pos in positions:
            # Players currently in the game and too tired
            tired_players = [p for p in self.defense if p.position == pos and p.in_game and p.fatigue > fatigue_threshold]
            if not tired_players:
                continue

            # Bench players eligible to sub in
            bench = [p for p in self.defense if p.position == pos and not p.in_game and p.fatigue <= fatigue_threshold]
            if not bench:
                continue

            # Prioritize by freshest and most productive
            def stat_sum(player):
                return sum(player.stats.values())

            bench.sort(key=lambda p: (p.fatigue, -stat_sum(p)))

            # Swap one-for-one
            for tired in tired_players:
                if not bench:
                    break
                fresh = bench.pop(0)
                tired.in_game = False
                fresh.in_game = True


    def get_kicker(self):
        for p in self.offense:
            if p.position == "K":
                return p
        return None
    
    def apply_fatigue_penalties(self):
        for player in self.offense + self.defense:
            fatigue_pct = player.fatigue / 100
            penalty_scale = fatigue_pct * 0.15  # max 15% drop

            for attr, original_value in player._original_attrs.items():
                if original_value is None:
                    continue

                min_allowed = round(original_value * (2 / 3))
                new_value = round(original_value * (1 - penalty_scale))
                setattr(player, attr, max(min_allowed, new_value))

def initialize_teams(json_path):
    with open(json_path, "r") as f:
        data = json.load(f)

    teams = []
    for team_data in data["teams"]:
        team = Team(
            name=team_data["team_name"],
            offense=team_data["offense"],
            defense=team_data["defense"]
        )
        teams.append(team)

    return teams

File: game_sim_ts/roster.ts
import { PlayerStats, Position, RawPlayerData, RawTeamData } from "./types";

// Player mirrors the Python Player class, including a snapshot of original
// attributes for fatigue penalties.
export class Player {
  name: string;
  position: Position;
  speed: number;
  strength: number;
  intelligence: number;
  endurance: number;
  fatigue: number;
  in_game: boolean;
  stats: PlayerStats;

  // Skill-specific
  passing?: number;
  decision_making?: number;
  elusiveness?: number;
  vision?: number;
  hands?: number;
  route_running?: number;
  run_blocking?: number;
  pass_blocking?: number;
  rushing?: number;
  tackling?: number;
  coverage?: number;
  kick_power?: number;
  kick_accuracy?: number;
  punt_power?: number;
  punt_accuracy?: number;

  private _originalAttrs: Partial<Record<string, number>>;

  constructor(data: RawPlayerData) {
    this.name = data.name;
    this.position = data.position;
    this.speed = data.speed ?? 50;
    this.strength = data.strength ?? 50;
    this.intelligence = data.intelligence ?? 50;
    this.endurance = data.endurance ?? 50;
    this.fatigue = data.fatigue ?? 0;
    this.in_game = data.in_game ?? false;
    this.stats = { ...(data.stats ?? {}) };

    this.passing = data.passing;
    this.decision_making = data.decision_making;
    this.elusiveness = data.elusiveness;
    this.vision = data.vision;
    this.hands = data.hands;
    this.route_running = data.route_running;
    this.run_blocking = data.run_blocking;
    this.pass_blocking = data.pass_blocking;
    this.rushing = data.rushing;
    this.tackling = data.tackling;
    this.coverage = data.coverage;
    this.kick_power = data.kick_power;
    this.kick_accuracy = data.kick_accuracy;
    this.punt_power = data.punt_power;
    this.punt_accuracy = data.punt_accuracy;

    this._originalAttrs = {
      speed: this.speed,
      strength: this.strength,
      elusiveness: this.elusiveness,
      vision: this.vision,
      hands: this.hands,
      route_running: this.route_running,
      run_blocking: this.run_blocking,
      pass_blocking: this.pass_blocking,
      rushing: this.rushing,
      tackling: this.tackling,
      coverage: this.coverage,
    };
  }

  is_offense(): boolean {
    return ["QB", "RB", "WR", "TE", "OL", "K", "P"].includes(this.position);
  }

  is_defense(): boolean {
    return ["DL", "ROLB", "MLB", "LOLB", "CB", "S"].includes(this.position);
  }

  fatigue_level(): number {
    return this.fatigue / Math.max(this.endurance, 1);
  }
}

export class Team {
  name: string;
  offense: Player[];
  defense: Player[];

  constructor(name: string, offense: RawPlayerData[], defense: RawPlayerData[]) {
    this.name = name;
    this.offense = offense.map((p) => new Player(p));
    this.defense = defense.map((p) => new Player(p));
  }

  get_offense(side: "offense" | "defense" = "offense"): Player[] {
    const players = (this as any)[side] as Player[];
    return players.filter((p) => p.in_game);
  }

  get_all_offense(side: "offense" | "defense" = "offense"): Player[] {
    const players = (this as any)[side] as Player[];
    return [...players];
  }

  get_defense(side: "offense" | "defense" = "defense"): Player[] {
    const players = (this as any)[side] as Player[];
    return players.filter((p) => p.in_game);
  }

  get_all_defense(side: "offense" | "defense" = "defense"): Player[] {
    const players = (this as any)[side] as Player[];
    return [...players];
  }

  get_player_by_name(name: string): Player | undefined {
    return [...this.offense, ...this.defense].find((p) => p.name === name);
  }

  recover_bench_players(): void {
    for (const p of [...this.offense, ...this.defense]) {
      if (!p.in_game) {
        const recovery = p.endurance / 10;
        p.fatigue = Math.max(0, p.fatigue - recovery);
      }
    }
  }

  sub_skill_position_players(fatigue_threshold = 50): void {
    const skillPositions: Position[] = ["RB", "WR", "TE"] as Position[];
    const statSum = (player: Player) =>
      Object.values(player.stats).reduce((a, b) => a + (b ?? 0), 0);

    for (const pos of skillPositions) {
      const starters = this.offense.filter(
        (p) => p.position === pos && p.in_game && p.fatigue > fatigue_threshold
      );
      if (!starters.length) continue;

      const bench = this.offense
        .filter((p) => p.position === pos && !p.in_game && p.fatigue <= fatigue_threshold)
        .sort((a, b) => {
          const statDiff = statSum(b) - statSum(a);
          if (statDiff !== 0) return statDiff;
          return a.fatigue - b.fatigue;
        });
      if (!bench.length) continue;

      for (const tired of starters) {
        const sub = bench.shift();
        if (!sub) break;
        tired.in_game = false;
        sub.in_game = true;
      }
    }
  }

  sub_defensive_players(fatigue_threshold = 50): void {
    const positions = new Set(this.defense.map((p) => p.position));
    const statSum = (player: Player) =>
      Object.values(player.stats).reduce((a, b) => a + (b ?? 0), 0);

    for (const pos of positions) {
      const tiredPlayers = this.defense.filter(
        (p) => p.position === pos && p.in_game && p.fatigue > fatigue_threshold
      );
      if (!tiredPlayers.length) continue;

      const bench = this.defense
        .filter((p) => p.position === pos && !p.in_game && p.fatigue <= fatigue_threshold)
        .sort((a, b) => {
          const fatigueDiff = a.fatigue - b.fatigue; // freshest first
          if (fatigueDiff !== 0) return fatigueDiff;
          return statSum(b) - statSum(a); // then more productive
        });
      if (!bench.length) continue;

      for (const tired of tiredPlayers) {
        const fresh = bench.shift();
        if (!fresh) break;
        tired.in_game = false;
        fresh.in_game = true;
      }
    }
  }

  get_kicker(): Player | undefined {
    return this.offense.find((p) => p.position === "K");
  }

  apply_fatigue_penalties(): void {
    for (const player of [...this.offense, ...this.defense]) {
      const fatiguePct = player.fatigue / 100;
      const penaltyScale = fatiguePct * 0.15; // max 15% drop

      const original = (player as any)._originalAttrs as Record<string, number | undefined>;
      for (const [attr, originalValue] of Object.entries(original)) {
        if (originalValue == null) continue;
        const minAllowed = Math.round(originalValue * (2 / 3));
        const newValue = Math.round(originalValue * (1 - penaltyScale));
        (player as any)[attr] = Math.max(minAllowed, newValue);
      }
    }
  }
}

export function initializeTeams(rawTeams: RawTeamData[]): Team[] {
  return rawTeams.map((t) => {
    const team = new Team(t.team_name, t.offense, t.defense);
    const anyOffenseActive = team.offense.some((p) => p.in_game);
    const anyDefenseActive = team.defense.some((p) => p.in_game);
    // Fallback: if no starters flagged in roster JSON, activate everyone
    if (!anyOffenseActive) team.offense.forEach((p) => (p.in_game = true));
    if (!anyDefenseActive) team.defense.forEach((p) => (p.in_game = true));
    return team;
  });
}


